### PR TITLE
[XboxOne] Added ability to support multiple gamepads

### PR DESCRIPTION
**Pull request by martinh-unity**
Can read gamepadId and userId from each gamepad.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs [hunk 1]
```diff
@@ -161,6 +208,76 @@ protected override void FinishSetup(InputDeviceBuilder builder)
...
+        public int xboxOneUserId
+        {
+            get
+            {
+                UpdatePadSettings();
+                return m_XboxOneUserId;
+            }
+        }
+
+        public new static List<XboxOneGamepad> all
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs]: No lists. Use array for s_Devices and ReadOnlyArray as return type here.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs [hunk 2]
```diff
@@ -161,6 +208,76 @@ protected override void FinishSetup(InputDeviceBuilder builder)
...
+
+        protected override void OnRemoved()
+        {
+            base.OnRemoved();
+
+            for (int i = 0; i < s_Devices.Count; i++)
+            {
+                if (s_Devices[i] != null && s_Devices[i].gamepadId == gamepadId)
+                {
+                    s_Devices.RemoveAt(i);
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs]: With `OnAdded` doing `Add(this)`, seems like this should just be `Remove(this)`. When switched to array, use ArrayHelpers.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs [hunk 3]
```diff
@@ -161,6 +208,76 @@ protected override void FinishSetup(InputDeviceBuilder builder)
...
+                return m_GamepadId;
+            }
+        }
+
+        public int xboxOneUserId
+        {
+            get
+            {
+                UpdatePadSettings();
+                return m_XboxOneUserId;
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs:226]: This should be part of the device configuration. There's no way with the current API to know when the user has changed other than to constantly poll the value. Configuration getters should retrieve the current configuration via a device command / IOCTL and then cache it. When the configuration changes, the backend should send a configuration change event.

**Review by Rene-Damm — CHANGES_REQUESTED** [review]

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs (hunk 1 above)
Comment (martinh-unity) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs]: Really? The problem is it doesn't know how many there might be. Also returning an array from a list just allocates more memory and creates garbage. We do try to minimize that on the consoles as the garbage collect is more expensive on consoles. However I'll sort out this change.

Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs]: You can use ArrayHelpers to allocate with capacity (same thing that list does internally). E.g. with a capacity of 10. With a list you get two heap allocations. For the price you pay for the List header object, you can easily pay for 5 extra entries in your array. And List.Append() is not guarantee at all to not create garbage. Internally, it just allocates with capacity as well (same thing you can do with ArrayHelpers).

Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs]: Also, no returning mutable internal structures from public APIs.

**Comment by Rene-Damm** [comment]
Please also add tests.

**Pull request merged**